### PR TITLE
Implement tags

### LINKED ISSUE #4: Group words by collection/tag
Currently the app is single-tenant, so there is no way to create different "groups" or "collections" of words.
There are at least two options for solving this:
- A list of word "collections"
- A list of tags for every word

In case of a tag-based solution, the existing MongoDB schema would have to be extended with a `tags` field, and the GraphQL query would have to be extended with a filter-by-tags argument. Both changes are purely additive and non-breaking for the frontend.

### PR DESCRIPTION
**Pull request by Yakimych**
Fixes https://github.com/MargaretKrutikova/words-api-graphql/issues/4
Non-breaking implementation for a list of tags per word.

### CODE REVIEW

File: lib/database/mongodb.ts [hunk 1]
```diff
@@ -29,15 +30,23 @@ export default (mPool: MongoClient): MongoDbService => {
...
-        .find()
+    getWords: async ({ page, itemsPerPage, tags }: PaginatedWordsRequest) => {
+      const getTagsFilter = () => {
+        if (tags === undefined) return {};
+        else if (tags.length === 0)
+          return { $or: [{ tags: undefined }, { tags: [] }] };
+        else return { tags: { $all: [...tags] } };
+      };
+
+      const filteredWordCollection = getWordsCollection().find(getTagsFilter());
```
Comment (MargaretKrutikova) [inline lib/database/mongodb.ts]: How about `filteredWordsCollection`? Or just `filteredWords`?

File: lib/database/mongodb.ts [hunk 2]
```diff
@@ -10,6 +10,7 @@ import { getReplacementFields } from "./util";
 export type PaginatedWordsRequest = {
   readonly page: number;
   readonly itemsPerPage: number;
+  readonly tags: ReadonlyArray<string> | undefined;
```
Comment (MargaretKrutikova) [inline lib/database/mongodb.ts]: Maybe `readonly tags?: ReadonlyArray<string>`? Or is it important that the field is present but can be `undefined`?

File: lib/database/mongodb.ts [hunk 3]
```diff
@@ -29,15 +30,23 @@ export default (mPool: MongoClient): MongoDbService => {
...
 
-    getWords: async ({ page, itemsPerPage }: PaginatedWordsRequest) => {
-      const getPaginatedPromise = getWordsCollection()
-        .find()
+    getWords: async ({ page, itemsPerPage, tags }: PaginatedWordsRequest) => {
+      const getTagsFilter = () => {
+        if (tags === undefined) return {};
+        else if (tags.length === 0)
+          return { $or: [{ tags: undefined }, { tags: [] }] };
+        else return { tags: { $all: [...tags] } };
```
Comment (MargaretKrutikova) [inline lib/database/mongodb.ts]: Is it the same as `return { tags: { $all: tags } }`? I don't think you need to spread unless there is some TS business going on there or you want to make sure you are creating a new array.

File: lib/database/mongodb.ts [hunk 4]
```diff
@@ -29,15 +30,23 @@ export default (mPool: MongoClient): MongoDbService => {
     getWord: (wordId: ObjectId) =>
       getWordsCollection().findOne({ _id: new ObjectId(wordId) }),
 
-    getWords: async ({ page, itemsPerPage }: PaginatedWordsRequest) => {
-      const getPaginatedPromise = getWordsCollection()
-        .find()
+    getWords: async ({ page, itemsPerPage, tags }: PaginatedWordsRequest) => {
+      const getTagsFilter = () => {
+        if (tags === undefined) return {};
+        else if (tags.length === 0)
```
Comment (MargaretKrutikova) [inline lib/database/mongodb.ts]: The formatting here is a bit funky, also you don't need `else` after `return` 🙂 
```typescript
if (tags === undefined) return {};
if (tags.length === 0) return { $or: [{ tags: undefined }, { tags: [] }] };

return { tags: { $all: [...tags] } };
```
I guess it is a matter of preference.

**Comment by MargaretKrutikova** [comment]
Shall we add some tags to the seed data? There was a file with initial words, do we need to add `swedish` tag and possibly even a tag for my personal collection?

File: lib/database/mongodb.ts (hunk 2 above)
Comment (Yakimych) [reply lib/database/mongodb.ts]: I might be mistaken, but I think `tags?: T` is just syntax sugar for `tags: T | undefined`. I can change it to `tags?` for conciseness.

File: lib/database/mongodb.ts (hunk 3 above)
Comment (Yakimych) [reply lib/database/mongodb.ts]: There was some TypeScript business going on indeed, I can double-check, but I had to spread to make it type-check.

**Comment by Yakimych** [comment]
> Shall we add some tags to the seed data? There was a file with initial words, do we need to add `swedish` tag and possibly even a tag for my personal collection?

Added to some entries.

File: lib/database/mongodb.ts (hunk 4 above)
Comment (Yakimych) [reply lib/database/mongodb.ts]: Removed the `else`s.

File: lib/database/mongodb.ts (hunk 3 above)
Comment (Yakimych) [reply lib/database/mongodb.ts]: Yep, it does not typecheck - we can take a look together, but I didn't dig too deep into this.

File: lib/database/mongodb.ts (hunk 2 above)
Comment (Yakimych) [reply lib/database/mongodb.ts]: Changed to `tags?`

File: lib/database/mongodb.ts (hunk 1 above)
Comment (Yakimych) [reply lib/database/mongodb.ts]: Changed to `filteredWordsCollection`.

File: lib/database/mongodb.ts (hunk 2 above)
Comment (MargaretKrutikova) [reply lib/database/mongodb.ts]: It is not the same actually, check this [TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKhDOwoF4oG8BQUoA8BcUiATgJYB2A5lAD5QCuZAJhAGbkSMYC+GGoksBMABMKdFlwB+AsXIVuvAMYB7MoigALCABttygnHWo0PFWqQB3ZUW2MDQ0ca5A)

**Comment by MargaretKrutikova** [comment]
Thanks a lot for your contribution!

**Review by MargaretKrutikova — APPROVED** [review]

**Pull request merged**